// @mantine/charts/esm/Heatmap/HeatmapWeeks.mjs
'use client';
import { jsx, Fragment } from 'react/jsx-runtime';
import { getHeatColor } from './get-heat-color/get-heat-color.mjs';

function HeatmapWeeks({
  data,
  datesRange,
  rectSize,
  gap,
  rectRadius,
  min,
  max,
  colors,
  withTooltip,
  setHoveredRect,
  getRectProps,
  getStyles
}) {
  const rectSizeWithGap = rectSize + gap;
  return /* @__PURE__ */ jsx(Fragment, { children: datesRange.map((week, weekIndex) => {
    const days = week.map((date, dayIndex) => {
      if (!date) {
        return null;
      }
      const hasValue = date in data && data[date] !== null;
      const rectValue = hasValue ? data[date] : null;
      return /* @__PURE__ */ jsx(
        "rect",
        {
          width: rectSize,
          height: rectSize,
          x: gap,
          y: dayIndex * rectSizeWithGap + gap,
          rx: rectRadius,
          "data-empty": !hasValue || void 0,
          fill: hasValue ? getHeatColor({ value: data[date], min, max, colors }) : void 0,
          onPointerEnter: withTooltip ? () => setHoveredRect({ date, value: rectValue }) : void 0,
          ...getRectProps?.({ date, value: rectValue }),
          ...getStyles("rect")
        },
        date
      );
    });
    return /* @__PURE__ */ jsx(
      "g",
      {
        transform: `translate(${weekIndex * rectSizeWithGap}, 0)`,
        "data-id": "week",
        children: days
      },
      weekIndex
    );
  }) });
}

export { HeatmapWeeks };
