// @mantine/charts/cjs/Heatmap/Heatmap.cjs
'use client';
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var react = require('react');
var core = require('@mantine/core');
var getBoundaries = require('./get-boundaries/get-boundaries.cjs');
var getColumns = require('./get-columns/get-columns.cjs');
var getDatesRange = require('./get-dates-range/get-dates-range.cjs');
var getMonthsRange = require('./get-months-range/get-months-range.cjs');
var HeatmapSplitWeeks = require('./HeatmapSplitWeeks.cjs');
var HeatmapWeeks = require('./HeatmapWeeks.cjs');
var rotateWeekdaysNames = require('./rotate-weekdays-names/rotate-weekdays-names.cjs');
var Heatmap_module = require('./Heatmap.module.css.cjs');

const defaultProps = {
  monthLabels: ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"],
  weekdayLabels: ["Sun", "Mon", "", "Wed", "", "Fri", ""],
  withOutsideDates: true,
  firstDayOfWeek: 1,
  rectSize: 10,
  weekdaysLabelsWidth: 30,
  monthsLabelsHeight: 14,
  gap: 1,
  rectRadius: 2,
  fontSize: 12,
  colors: [
    "var(--heatmap-level-1)",
    "var(--heatmap-level-2)",
    "var(--heatmap-level-3)",
    "var(--heatmap-level-4)"
  ]
};
const Heatmap = core.factory((_props, ref) => {
  const props = core.useProps("Heatmap", defaultProps, _props);
  const {
    classNames,
    className,
    style,
    styles,
    unstyled,
    vars,
    data,
    startDate,
    endDate,
    withMonthLabels,
    withWeekdayLabels,
    weekdayLabels,
    withOutsideDates,
    monthLabels,
    firstDayOfWeek,
    rectSize = 10,
    gap = 1,
    rectRadius,
    domain,
    colors,
    weekdaysLabelsWidth,
    monthsLabelsHeight,
    fontSize,
    getTooltipLabel,
    withTooltip,
    tooltipProps,
    getRectProps,
    splitMonths,
    attributes,
    ...others
  } = props;
  const getStyles = core.useStyles({
    name: "Heatmap",
    classes: Heatmap_module,
    props,
    className,
    style,
    classNames,
    styles,
    unstyled,
    attributes,
    vars
  });
  const [hoveredRect, setHoveredRect] = react.useState(null);
  const rectSizeWithGap = rectSize + gap;
  const weekdaysOffset = withWeekdayLabels ? weekdaysLabelsWidth : 0;
  const monthsOffset = withMonthLabels ? monthsLabelsHeight : 0;
  const [min, max] = getBoundaries.getBoundaries({ data, domain });
  const rotatedWeekdayLabels = react.useMemo(
    () => rotateWeekdaysNames.rotateWeekdaysNames(weekdayLabels, firstDayOfWeek),
    [weekdayLabels, firstDayOfWeek]
  );
  const datesRange = getDatesRange.getDatesRange({
    startDate,
    endDate,
    withOutsideDates,
    firstDayOfWeek
  });
  const monthsRange = withMonthLabels || splitMonths ? getMonthsRange.getMonthsRange(datesRange) : [];
  const weeksProps = {
    data,
    datesRange,
    rectSize,
    gap,
    rectRadius,
    min,
    max,
    colors,
    withTooltip,
    setHoveredRect,
    getRectProps,
    getStyles
  };
  const weeks = splitMonths ? /* @__PURE__ */ jsxRuntime.jsx(HeatmapSplitWeeks.HeatmapSplitWeeks, { ...weeksProps }) : /* @__PURE__ */ jsxRuntime.jsx(HeatmapWeeks.HeatmapWeeks, { ...weeksProps });
  const totalColumns = splitMonths ? getColumns.getColumns(datesRange, splitMonths).length : datesRange.length;
  const computeMonthLabelX = (monthPosition, monthIndex) => {
    if (!splitMonths) {
      return monthPosition * rectSizeWithGap + gap + weekdaysOffset;
    }
    const firstMonth = monthsRange[monthIndex];
    const columns = getColumns.getColumns(datesRange, splitMonths);
    const i = getColumns.getFirstMonthColumnIndex(columns, firstMonth.month);
    const base = i >= 0 ? i : monthPosition;
    return (base + 1) * rectSizeWithGap + gap + weekdaysOffset;
  };
  const monthsLabelsNodes = withMonthLabels && monthLabels ? monthsRange.map((month, monthIndex) => {
    const minSize = splitMonths ? 2 : 3;
    if (month.size < minSize) {
      return null;
    }
    const monthLabel = monthLabels[month.month];
    return /* @__PURE__ */ jsxRuntime.jsx(
      "text",
      {
        x: computeMonthLabelX(month.position, monthIndex),
        y: monthsLabelsHeight - 4,
        width: month.size * rectSizeWithGap,
        fontSize,
        ...getStyles("monthLabel"),
        children: monthLabel
      },
      monthIndex
    );
  }) : null;
  const weekdayLabelsNodes = withWeekdayLabels && weekdayLabels ? rotatedWeekdayLabels.map((weekdayLabel, dayIndex) => /* @__PURE__ */ jsxRuntime.jsx(
    "text",
    {
      x: 0,
      y: (dayIndex + 1) * rectSizeWithGap - gap + monthsOffset,
      width: weekdaysLabelsWidth,
      fontSize,
      ...getStyles("weekdayLabel"),
      children: weekdayLabel
    },
    dayIndex
  )) : null;
  const label = getTooltipLabel && hoveredRect && withTooltip ? getTooltipLabel(hoveredRect) : null;
  return /* @__PURE__ */ jsxRuntime.jsxs(
    core.Box,
    {
      component: "svg",
      ref,
      width: rectSizeWithGap * totalColumns + gap + weekdaysOffset,
      height: rectSizeWithGap * 7 + gap + monthsOffset,
      ...getStyles("root"),
      ...others,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          core.Tooltip.Floating,
          {
            label,
            disabled: !withTooltip || !label,
            position: "top",
            ...tooltipProps,
            children: /* @__PURE__ */ jsxRuntime.jsxs("g", { transform: `translate(${weekdaysOffset}, ${monthsOffset})`, "data-id": "all-weeks", children: [
              withTooltip && /* @__PURE__ */ jsxRuntime.jsx(
                "rect",
                {
                  fill: "transparent",
                  width: rectSizeWithGap * totalColumns + gap,
                  height: rectSizeWithGap * 7 + gap
                }
              ),
              weeks
            ] })
          }
        ),
        weekdayLabelsNodes,
        monthsLabelsNodes
      ]
    }
  );
});
Heatmap.displayName = "@mantine/charts/Heatmap";
Heatmap.classes = Heatmap_module;

exports.Heatmap = Heatmap;
